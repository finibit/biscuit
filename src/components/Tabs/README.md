### Usage

#### Items

```jsx
class CustomTabs extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      activeIndex: 1,
    }

    this.onTabChange = this.onTabChange.bind(this)
  }

  onTabChange(event, index) {
    this.setState({
      activeIndex: index,
    })
  }

  render() {
    return (
      <Tabs activeIndex={this.state.activeIndex} onTabChange={this.onTabChange}>
        <Tabs.Item title="Tab 1" />
        <Tabs.Item title="Tab 2" />
        <Tabs.Item title="Tab 3" disabled />
        <Tabs.Item title="Tab 4" disabled hidden />
      </Tabs>
    )
  }
}

<CustomTabs />
```

