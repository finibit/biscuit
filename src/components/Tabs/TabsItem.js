import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { themePadding, responsiveArray } from '../../themes/themeUtils'

const TabsItemStyled = styled.div`
	box-sizing: border-box;

	margin: 0;
	padding: ${(props) => themePadding(props.theme, 'tabs', props.$padding[0])};

	@media only screen and (min-width: ${(props) => props.theme.global.breakpoints.sm}) {
		padding: ${(props) => themePadding(props.theme, 'tabs', props.$padding[1])};
	}

	@media only screen and (min-width: ${(props) => props.theme.global.breakpoints.md}) {
		padding: ${(props) => themePadding(props.theme, 'tabs', props.$padding[2])};
	}

	@media only screen and (min-width: ${(props) => props.theme.global.breakpoints.lg}) {
		padding: ${(props) => themePadding(props.theme, 'tabs', props.$padding[3])};
	}
`

const TabsItem = (props) => {
	const {
		padding,
		...rest
	} = props

	return (
		<TabsItemStyled
			$padding={responsiveArray(padding)}
			{...rest}
		/>
	)
}

TabsItem.propTypes = {
	/** Any number of renderable nodes. */
	children: PropTypes.node,

	/** The title of the tab. */
	title: PropTypes.string.isRequired,

	/** Is this tab disabled? */
	disabled: PropTypes.bool,

	/** The amount of padding around the tab content. */
	padding: PropTypes.oneOfType([PropTypes.string, PropTypes.object, PropTypes.array]),
}

TabsItem.defaultProps = {
	children: null,
	disabled: false,
	padding: { horizontal: 'none', vertical: 'md' },
}

export default TabsItem
