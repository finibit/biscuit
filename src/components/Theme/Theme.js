import React from 'react'
import PropTypes from 'prop-types'
import { ThemeProvider, createGlobalStyle } from 'styled-components'
import { mergeTheme, baseTheme, themeGet } from '../../themes'
import types from '../../types'

const GlobalStyle = createGlobalStyle`
	html, body {
		width: 100%;
		height: 100%;
		margin: 0;
		padding: 0;
	}

	body {
		color: ${themeGet.color('black', 'inherit')};
		font-family: ${themeGet.fontFamily('secondary')};
		-webkit-font-smoothing: antialiased;
	}
`

const Theme = (props) => {
	const {
		theme,
		global,
	} = props

	return (
		<ThemeProvider theme={mergeTheme(baseTheme, theme)}>
			<div>
				{ global && <GlobalStyle /> }
				{props.children}
			</div>
		</ThemeProvider>
	)
}

Theme.propTypes = {
	global: PropTypes.bool,
	...types.theme,
}

Theme.defaultProps = {
	global: true,
	theme: baseTheme,
}

export default Theme
