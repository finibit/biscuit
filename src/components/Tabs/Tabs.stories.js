import React, { useState } from 'react'
import { storiesOf } from '@storybook/react'
import Wrapper from '../../Wrapper'
import Tabs from './Tabs'

const TabsWithState = (props) => {
  const [activeIndex, setActiveIndex] = useState(0)

  return (
    <Tabs
      activeIndex={activeIndex}
      onTabChange={(event, index) => setActiveIndex(index)}
      {...props}
    >
      {props.children}
    </Tabs>
  )
}

storiesOf('Tabs', module)
  .addDecorator(Wrapper)
  .add('with items', () => (
    <TabsWithState>
      <Tabs.Item title='Tab 1'>
        Content of Tab 1
      </Tabs.Item>
      <Tabs.Item title='Tab 2'>
        Content of Tab 2
      </Tabs.Item>
      <Tabs.Item title='Tab 3'>
        Content of Tab 3
      </Tabs.Item>
      <Tabs.Item title='Tab 4' disabled>
        Content of Tab 4
      </Tabs.Item>
      <Tabs.Item title='Tab 5' hidden>
        Content of Tab 5
      </Tabs.Item>
    </TabsWithState>
  ))
