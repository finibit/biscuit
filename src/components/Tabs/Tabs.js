import React, { Children } from 'react'
import PropTypes from 'prop-types'
import styled, { css } from 'styled-components'
import TabsItem from './TabsItem'
import styles, { utils } from '../../themes'

const TabsStyled = styled.div`
	width: 100%;
	box-sizing: border-box;
	margin: 0;
	padding: 0;
`

const TabsNavStyled = styled.ul`
	margin: 0;
	padding: 0;
	width: 100%;
	list-style: none;
	display: flex;
	flex-direction: row;
	flex-wrap: nowrap;
	justify-content: flext-start;
	border-bottom: 2px solid ${(props) => utils.resolveColor(props.theme, props.$element, 'inactiveTabBorder')};
	box-sizing: border-box;
`

const TabsNavItemStyled = styled.li`
	margin: 0;
	padding: 0;
	box-sizing: border-box;
	white-space: nowrap;
`

const inactiveStyles = css`
	border-bottom: 2px solid transparent;

	&:hover, &:active {
		border-bottom: 2px solid ${(props) => utils.resolveColor(props.theme, props.$element, 'hoverTabBorder')};
	}
`

const activeStyles = css`
	border-bottom: 2px solid ${(props) => utils.resolveColor(props.theme, props.$element, 'activeTabBorder')};

	&:hover, &:active {
		border-bottom: 2px solid ${(props) => utils.resolveColor(props.theme, props.$element, 'activeTabBorder')};
	}
`

const disabledStyles = css`
	border-bottom: 2px solid transparent;

	&:hover, &:active {
		border-bottom: 2px solid ${(props) => utils.resolveColor(props.theme, props.$element, 'disabledTabBorder')};
	}

	color: ${(props) => utils.resolveColor(props.theme, props.$element, 'disabledTabTitle')} !important;
`

const TabsButtonStyled = styled.a`
	${styles.fontFamily}
	${styles.padding}

	color: ${(props) => props.theme.global.colors['dark-0']};
	display: inline-block;
	cursor: pointer;
	font-size: 1rem;
	font-weight: normal;
	text-decoration: none;
	box-sizing: border-box;
	line-height: 1;
	vertical-align: middle;

	margin: 0 0 -2px 0;

	${(props) => {
		if (props.$disabled) {
			return disabledStyles
		}

		if (props.$active) {
			return activeStyles
		}

		return inactiveStyles
	}};
`

const TabsButton = (props) => {
	const {
		index,
		active,
		disabled,
		padding,
		children,
		onClick,
		...rest
	} = props

	return (
		<TabsButtonStyled
			$element="Tabs"
			$active={active}
			$disabled={disabled}
			$padding={padding}
			{...rest}
			onClick={(event) => {
				if (!disabled) {
					event.preventDefault()
					onClick(event, index)
				}
			}}
		>
			{children}
		</TabsButtonStyled>
	)
}

TabsButton.propTypes = {
	children: PropTypes.string.isRequired,
	index: PropTypes.number.isRequired,
	active: PropTypes.bool.isRequired,
	disabled: PropTypes.bool.isRequired,
	padding: PropTypes.oneOfType([
		PropTypes.number,
		PropTypes.string,
		PropTypes.object,
		PropTypes.array,
	]),
	onClick: PropTypes.func,
}

TabsButton.defaultProps = {
	padding: 'none',
	onClick: () => {},
}

const Tabs = (props) => {
	const {
		children,
		padding,
		activeIndex,
		onTabChange,
		themeElement,
		...rest
	} = props

	return (
		<TabsStyled
			$element={themeElement}
			{...rest}
		>
			<TabsNavStyled
				$element={themeElement}
			>
				{Children.map(children, (child, index) => (
					<TabsNavItemStyled
						key={child.props.title}
						$element={themeElement}
					>
						<TabsButton
							$element={themeElement}
							index={index}
							active={index === activeIndex}
							disabled={child.props.disabled}
							padding={padding}
							onClick={(event) => {
								onTabChange(event, index)
							}}
						>
							{child.props.title}
						</TabsButton>
					</TabsNavItemStyled>
				))}
			</TabsNavStyled>
			{Children.toArray(props.children)[activeIndex]}
		</TabsStyled>
	)
}

Tabs.propTypes = {
	/** At least one `Tabs.Item` components. */
	children: PropTypes.node.isRequired,

	/** The amount of padding around the tab navigation. */
	padding: PropTypes.oneOfType([
		PropTypes.number,
		PropTypes.string,
		PropTypes.object,
		PropTypes.array,
	]),

	/** Active item index. */
	activeIndex: PropTypes.number.isRequired,

	/**
	 * Called when tab change is requested.
	 *
	 * @param {SyntheticEvent} event - React's original SyntheticEvent
	 * @param {number} index - Proposed new activeIndex
	 */
	onTabChange: PropTypes.func.isRequired,

	/** Theme element. */
	themeElement: PropTypes.string,
}

Tabs.defaultProps = {
	padding: { horizontal: 2, vertical: 1 },
	themeElement: 'Tabs',
}

Tabs.Item = TabsItem

export default Tabs
