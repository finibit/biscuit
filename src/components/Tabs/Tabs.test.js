import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import Theme from '../Theme'
import Tabs from '../Tabs'

describe('Tabs', () => {
	test('navigation is rendered', () => {
		const component = renderer.create((
			<Theme>
				<Tabs
					activeIndex={1}
					onTabChange={() => {}}
				>
					<Tabs.Item title="" />
					<Tabs.Item title="" />
				</Tabs>
			</Theme>
		))
		expect(component.toJSON()).toMatchSnapshot()
	})

	test('navigation is rendered with custom padding', () => {
		const component = renderer.create((
			<Theme>
				<Tabs
					activeIndex={0}
					onTabChange={() => {}}
					padding="xl"
				>
					<Tabs.Item title="" />
				</Tabs>
			</Theme>
		))
		expect(component.toJSON()).toMatchSnapshot()
	})

	test('disabled navigation item is rendered', () => {
		const component = renderer.create((
			<Theme>
				<Tabs
					activeIndex={0}
					onTabChange={() => {}}
				>
					<Tabs.Item title="" />
					<Tabs.Item title="" disabled />
				</Tabs>
			</Theme>
		))
		expect(component.toJSON()).toMatchSnapshot()
	})
})

describe('TabsItem', () => {
	test('is rendered', () => {
		const component = renderer.create((
			<Theme>
				<Tabs.Item title="Item" />
			</Theme>
		))
		expect(component.toJSON()).toMatchSnapshot()
	})

	test('is rendered with custom padding', () => {
		const component = renderer.create((
			<Theme>
				<Tabs.Item title="Item" padding="none" />
				<Tabs.Item title="Item" padding="1rem" />
				<Tabs.Item title="Item" padding="xs" />
				<Tabs.Item title="Item" padding={{ horizontal: 'xs', vertical: 'sm' }} />
				<Tabs.Item title="Item" padding={{ top: 'xs', left: 'sm', bottom: 'lg', right: 'xl' }} />
				<Tabs.Item title="Item" padding={['xs', 'sm', 'lg', 'xl']} />
			</Theme>
		))
		expect(component.toJSON()).toMatchSnapshot()
	})
})
