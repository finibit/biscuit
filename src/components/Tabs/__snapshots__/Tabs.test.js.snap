// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs disabled navigation item is rendered 1`] = `
.c5 {
  margin: 0;
  padding: 1rem 0 1rem 0;
  box-sizing: border-box;
}

.c0 {
  width: 100%;
  box-sizing: border-box;
  margin: 0;
  padding: 0;
}

.c1 {
  margin: 0;
  padding: 0;
  width: 100%;
  list-style: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: flext-start;
  -webkit-justify-content: flext-start;
  -ms-flex-pack: flext-start;
  justify-content: flext-start;
  border-bottom: 2px solid #e5e5e5;
  box-sizing: border-box;
}

.c2 {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
  white-space: nowrap;
}

.c3 {
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif;
  padding: 0.694rem 1rem 0.694rem 1rem;
  display: inline-block;
  cursor: pointer;
  font-size: 1rem;
  font-weight: normal;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-sizing: border-box;
  line-height: 1;
  vertical-align: middle;
  margin: 0 0 -2px 0;
  border-bottom: 2px solid #000000;
}

.c3:hover,
.c3:active {
  border-bottom: 2px solid #000000;
}

.c4 {
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif;
  padding: 0.694rem 1rem 0.694rem 1rem;
  display: inline-block;
  cursor: pointer;
  font-size: 1rem;
  font-weight: normal;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-sizing: border-box;
  line-height: 1;
  vertical-align: middle;
  margin: 0 0 -2px 0;
  border-bottom: 2px solid transparent;
  color: #ccc !important;
}

.c4:hover,
.c4:active {
  border-bottom: 2px solid #e5e5e5;
}

<div>
  <div
    className="c0"
  >
    <ul
      className="c1"
    >
      <li
        className="c2"
      >
        <a
          className="c3"
          onClick={[Function]}
        >
          
        </a>
      </li>
      <li
        className="c2"
      >
        <a
          className="c4"
          onClick={[Function]}
        >
          
        </a>
      </li>
    </ul>
    <div
      className="c5"
      disabled={false}
      title=""
    />
  </div>
</div>
`;

exports[`Tabs navigation is rendered 1`] = `
.c5 {
  margin: 0;
  padding: 1rem 0 1rem 0;
  box-sizing: border-box;
}

.c0 {
  width: 100%;
  box-sizing: border-box;
  margin: 0;
  padding: 0;
}

.c1 {
  margin: 0;
  padding: 0;
  width: 100%;
  list-style: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: flext-start;
  -webkit-justify-content: flext-start;
  -ms-flex-pack: flext-start;
  justify-content: flext-start;
  border-bottom: 2px solid #e5e5e5;
  box-sizing: border-box;
}

.c2 {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
  white-space: nowrap;
}

.c3 {
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif;
  padding: 0.694rem 1rem 0.694rem 1rem;
  display: inline-block;
  cursor: pointer;
  font-size: 1rem;
  font-weight: normal;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-sizing: border-box;
  line-height: 1;
  vertical-align: middle;
  margin: 0 0 -2px 0;
  border-bottom: 2px solid transparent;
}

.c3:hover,
.c3:active {
  border-bottom: 2px solid #ccc;
}

.c4 {
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif;
  padding: 0.694rem 1rem 0.694rem 1rem;
  display: inline-block;
  cursor: pointer;
  font-size: 1rem;
  font-weight: normal;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-sizing: border-box;
  line-height: 1;
  vertical-align: middle;
  margin: 0 0 -2px 0;
  border-bottom: 2px solid #000000;
}

.c4:hover,
.c4:active {
  border-bottom: 2px solid #000000;
}

<div>
  <div
    className="c0"
  >
    <ul
      className="c1"
    >
      <li
        className="c2"
      >
        <a
          className="c3"
          onClick={[Function]}
        >
          
        </a>
      </li>
      <li
        className="c2"
      >
        <a
          className="c4"
          onClick={[Function]}
        >
          
        </a>
      </li>
    </ul>
    <div
      className="c5"
      disabled={false}
      title=""
    />
  </div>
</div>
`;

exports[`Tabs navigation is rendered with custom padding 1`] = `
.c4 {
  margin: 0;
  padding: 1rem 0 1rem 0;
  box-sizing: border-box;
}

.c0 {
  width: 100%;
  box-sizing: border-box;
  margin: 0;
  padding: 0;
}

.c1 {
  margin: 0;
  padding: 0;
  width: 100%;
  list-style: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: flext-start;
  -webkit-justify-content: flext-start;
  -ms-flex-pack: flext-start;
  justify-content: flext-start;
  border-bottom: 2px solid #e5e5e5;
  box-sizing: border-box;
}

.c2 {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
  white-space: nowrap;
}

.c3 {
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif;
  padding: xl;
  display: inline-block;
  cursor: pointer;
  font-size: 1rem;
  font-weight: normal;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-sizing: border-box;
  line-height: 1;
  vertical-align: middle;
  margin: 0 0 -2px 0;
  border-bottom: 2px solid #000000;
}

.c3:hover,
.c3:active {
  border-bottom: 2px solid #000000;
}

<div>
  <div
    className="c0"
  >
    <ul
      className="c1"
    >
      <li
        className="c2"
      >
        <a
          className="c3"
          onClick={[Function]}
        >
          
        </a>
      </li>
    </ul>
    <div
      className="c4"
      disabled={false}
      title=""
    />
  </div>
</div>
`;

exports[`TabsItem is rendered 1`] = `
.c0 {
  margin: 0;
  padding: 1rem 0 1rem 0;
  box-sizing: border-box;
}

<div>
  <div
    className="c0"
    disabled={false}
    title="Item"
  />
</div>
`;

exports[`TabsItem is rendered with custom padding 1`] = `
.c0 {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

.c1 {
  margin: 0;
  padding: 1rem;
  box-sizing: border-box;
}

.c2 {
  margin: 0;
  padding: xs;
  box-sizing: border-box;
}

.c3 {
  margin: 0;
  padding: smrem xsrem smrem xsrem;
  box-sizing: border-box;
}

.c4 {
  margin: 0;
  padding: xsrem xlrem lgrem smrem;
  box-sizing: border-box;
}

.c5 {
  margin: 0;
  padding: xs;
  box-sizing: border-box;
}

@media only screen and (min-width:48rem) {
  .c5 {
    padding: sm;
  }
}

@media only screen and (min-width:64rem) {
  .c5 {
    padding: lg;
  }
}

@media only screen and (min-width:75rem) {
  .c5 {
    padding: xl;
  }
}

<div>
  <div
    className="c0"
    disabled={false}
    title="Item"
  />
  <div
    className="c1"
    disabled={false}
    title="Item"
  />
  <div
    className="c2"
    disabled={false}
    title="Item"
  />
  <div
    className="c3"
    disabled={false}
    title="Item"
  />
  <div
    className="c4"
    disabled={false}
    title="Item"
  />
  <div
    className="c5"
    disabled={false}
    title="Item"
  />
</div>
`;
